### PR TITLE
[CELEBORN-1401] Add SSL support for ratis communication

### PR DESCRIPTION
**Pull request by mridulm**
### What changes were proposed in this pull request?

When SSL is enabled for master, secure the Ratis communication as well with TLS

### Why are the changes needed?

Currently, when TLS is enabled for RPC, Ratis comms still goes in the clear - add support for TLS.
Note that currently this only supports GRPC, and not netty.

### Does this PR introduce _any_ user-facing change?
Secures ratis communication when TLS is enabled at master for rpc.

### How was this patch tested?
Local tests and additional unit tests added

### CODE REVIEW

File: master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/ha/HARaftServer.java [hunk 1]
```diff
@@ -375,6 +386,37 @@ private RaftProperties newRaftProperties(CelebornConf conf) {
     return properties;
   }
 
+  private Parameters configureSsl(CelebornConf conf, RaftProperties properties, RpcType rpc) {
+
+    if (rpc != SupportedRpcType.GRPC) {
+      LOG.error(
+          "SSL has been disabled for Raft communication between masters. "
+              + "This is only supported when ratis is configured with GRPC");
```
Comment (mridulm) [inline master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/ha/HARaftServer.java:394]: QQ: The default is NETTY - are there any concerns with changing default to GRPC ?

**Comment by mridulm** [comment]
+CC @otterc, @waitinfuture , @FMX , @SteNicholas 
PTAL, thanks !

File: master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/ha/HARaftServer.java (hunk 1 above)
Comment (RexXiong) [reply master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/ha/HARaftServer.java:394]: If a rolling upgrade of the master is performed, might there be compatibility issues with Netty RPC here?

**Review by RexXiong — APPROVED** [review]
LGTM

File: master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/ha/HARaftServer.java (hunk 1 above)
Comment (mridulm) [reply master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/ha/HARaftServer.java:394]: That is a good point, agree - thanks !
Let us leave it as is, I have documented it explicitly to use GRPC until Ratis (and so we) add support for netty.

File: master/src/test/java/org/apache/celeborn/service/deploy/master/clustermeta/ha/SSLRatisMasterStatusSystemSuiteJ.java [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+      socket.setSoTimeout(5000);
+      socket.startHandshake();
+      // handshake succeeded, this will always return true in this case
+      return socket.getSession().isValid();
+    }
+  }
+
+  private static SSLSocket createSslSocket(InetAddress address, int port) throws Exception {
+    TrustManager trustStore = SSLFactory.defaultTrustManagers(caData.get().file, "password")[0];
+    SSLContext context = SSLContext.getInstance("TLS");
```
Comment (byteroll) [inline master/src/test/java/org/apache/celeborn/service/deploy/master/clustermeta/ha/SSLRatisMasterStatusSystemSuiteJ.java:151]: If I change the `TLS` in this line to `TLSv1.2`, the unit test will fail. But if I change it to `TLSv1.3`, it runs correctly. What puzzles me is that when the server starts, the configuration is set to `TLSv1.2`, so why does the client report an error when it's also set to `TLSv1.2`? Any ideas?

Comment (byteroll) [reply master/src/test/java/org/apache/celeborn/service/deploy/master/clustermeta/ha/SSLRatisMasterStatusSystemSuiteJ.java:151]: the exception stack
```
java.net.SocketException: Broken pipe (Write failed)

	at java.net.SocketOutputStream.socketWrite0(Native Method)
	at java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:111)
	at java.net.SocketOutputStream.write(SocketOutputStream.java:155)
	at sun.security.ssl.SSLSocketOutputRecord.encodeChangeCipherSpec(SSLSocketOutputRecord.java:233)
	at sun.security.ssl.OutputRecord.changeWriteCiphers(OutputRecord.java:188)
	at sun.security.ssl.ChangeCipherSpec$T10ChangeCipherSpecProducer.produce(ChangeCipherSpec.java:118)
	at sun.security.ssl.Finished$T12FinishedProducer.onProduceFinished(Finished.java:390)
	at sun.security.ssl.Finished$T12FinishedProducer.produce(Finished.java:374)
	at sun.security.ssl.SSLHandshake.produce(SSLHandshake.java:420)
	at sun.security.ssl.ServerHelloDone$ServerHelloDoneConsumer.consume(ServerHelloDone.java:182)
	at sun.security.ssl.SSLHandshake.consume(SSLHandshake.java:376)
	at sun.security.ssl.HandshakeContext.dispatch(HandshakeContext.java:479)
	at sun.security.ssl.HandshakeContext.dispatch(HandshakeContext.java:457)
	at sun.security.ssl.TransportContext.dispatch(TransportContext.java:200)
	at sun.security.ssl.SSLTransport.decode(SSLTransport.java:155)
	at sun.security.ssl.SSLSocketImpl.decode(SSLSocketImpl.java:1320)
	at sun.security.ssl.SSLSocketImpl.readHandshakeRecord(SSLSocketImpl.java:1233)
	at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:417)
	at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:389)
	at org.apache.celeborn.service.deploy.master.clustermeta.ha.SSLRatisMasterStatusSystemSuiteJ.isSslServer(SSLRatisMasterStatusSystemSuiteJ.java:141)
```

**Pull request closed without merging**